src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlockMeasureCache.MeasureEntry.cs

```diff
@@ -11,66 +11,121 @@ internal partial class TextBlockMeasureCache
+				var currentAvailableHeight = availableSize.Height;
+
+				if(_sizes.TryGetValue(CachedTuple.Create(currentAvailableWidth, currentAvailableHeight), out var sizeEntry))
 				{
 					return sizeEntry.MeasuredSize;
 				}
-				else
+
+				var isWrapping = key.IsWrapping;
+
+				foreach (var kvp in _sizes)
 				{
-					if(key.TextWrapping == TextWrapping.NoWrap)
+					var size = kvp.Key;
+					var measureSizeEntry = kvp.Value;
+					var measurementCachedWidth = measureSizeEntry.MeasuredSize.Width;
+					var measurementCachedHeight = measureSizeEntry.MeasuredSize.Height;
+					var measurementAvailableWidth = size.Item1;
+					var measurementAvailableHeight = size.Item2;
+
+					if (isWrapping)
+					{
+						if (measurementCachedWidth <= currentAvailableWidth
+							&& currentAvailableWidth <= measurementAvailableWidth)
+						{
+							// Ok we can reuse it
+						}
+						else
+						{
+							continue; // Check for another cached measurement
+						}
+					}
+					else
 					{
-						// No wrap, assume any measured width below the asked available size
-						// is valid, if the available size is greater.
-						foreach(var keySize in _sizes)
+						// Non-wrapping text
+
+						if (double.IsInfinity(measurementAvailableWidth))
+						{
+							// Previous measurement was unconstrained
+							// horizontally: we can definitely reuse it.
+						}
+						else
 						{
-							if(keySize.Key.Item1 >= availableSize.Width &&
-								keySize.Value.MeasuredSize.Width <= availableSize.Width)
+							if (Math.Abs(measurementCachedWidth - measurementAvailableWidth) < 0.5d)
 							{
-								MoveToLast(keySize.Key, keySize.Value);
-								return keySize.Value.MeasuredSize;
+								// This measure was constrained, we can reuse only if the width is the same.
+
+								if (Math.Abs(measurementCachedWidth - currentAvailableWidth) < 0.5d)
+								{
+									// Yep, that's good
+								}
+
+								else
+								{
+									continue; // Check for another cached measurement
+								}
 							}
 						}
 					}
+
+					// We need to make sure the height is ok
+					if (double.IsInfinity(measurementAvailableHeight))
+					{
+						// Previous measurement was unconstrained
+						// vertically: we can definitely reuse it.
+					}
 					else
 					{
-						foreach (var keySize in _sizes)
+						// A max-height was specified in the cached measurement:
+						// We must check if we can reuse it.
+						if (Math.Abs(measurementCachedHeight - measurementAvailableHeight) < 0.5d)
 						{
-							// If text wraps and the available width is the same, any height below the
-							// available size is valid.
-							if (
-								keySize.Key.Item1 == availableSize.Width
-								&& keySize.Value.MeasuredSize.Height <= availableSize.Height
-							)
+							// This measure was constrained, we can reuse only if the available height
+							// is same or higher than current available height
+							if (measurementCachedHeight >= currentAvailableHeight)
```

Comment:
Same thing here : one `if` with all the conditions?

Reply:
Same comment...